include/slang/binding/OperatorExpressions.h

```diff
@@ -248,20 +248,35 @@ class ReplicationExpression : public Expression {
 
 struct StreamingConcatenationExpressionSyntax;
 
+#define RANGE(x) x(Simple) x(IndexedUp) x(IndexedDown) x(Bit)
+ENUM(WithRangeKind, RANGE); // RangeSelectionKind + Bit
+#undef RANGE
+
 /// Represents a streaming concatenation
 class StreamingConcatenationExpression : public Expression {
 public:
     /// The size of the blocks to slice and reorder: if 0, this is a left-to-right
     /// concatenation. Otherwise, it's a right-to-left concatenation.
     const size_t sliceSize;
 
+    struct WithExpression {
+        WithRangeKind kind;
+        Expression* left;
```

Comment:
I think left is always non-null, so can you use the not_null<> template to indicate that?

Reply:
Also can these be stored as const Expressions?

Reply:
Yes, I add not_null and const where appropriate.
```c++
    struct WithExpression {
        WithRangeKind kind;
        not_null<const Expression*> left;
        const Expression* right;
        optional<int32_t> width; // elaboration-time constant width
    };
    struct StreamExpression {
        not_null<const Expression*> operand;
        const WithExpression* with;
    };
```